### PR TITLE
added much more robust seed management utilities

### PR DESCRIPTION
**Pull request by josephdviviano**
- [x] I've read the [.github/CONTRIBUTING.md](.github/CONTRIBUTING.md) file
- [x] My code follows the typing guidelines
- [x] I've added appropriate tests
- [x] I've run pre-commit hooks locally

## Description
Adds a much more robust set of seeding utilities that operates across various hardware configurations. This has resolved reproducibility issues we observed with large scale multi-node runs.

### CODE REVIEW

**Comment by josephdviviano** [comment]
This fix was tested to fix run-to-run variation here https://github.com/GFNOrg/torchgfn/issues/307#issuecomment-3027954815

File: src/gfn/utils/common.py [hunk 1]
```diff
@@ -1,23 +1,172 @@
...
 import random
+import threading
+import warnings
+from typing import Callable, Tuple
 
 import numpy as np
 import torch
+import torch.distributed as dist
+
+# -----------------------------------------------------------------------------
```
Comment (younik) [inline src/gfn/utils/common.py:11]: the changes in common.py are helpful in a distributed settings. Should they go in a distributed module (just a file in utils called `distributed.py`)?

Comment (josephdviviano) [reply src/gfn/utils/common.py:11]: I'm not sure I totally understand this one - the logic here is `if distributed: do distributed things, else: do non-distributed things` -- it means the user can just `set_seed(number)` and it will work regardless if the code is being run in the distributed or non-distributed context.

Comment (younik) [reply src/gfn/utils/common.py:11]: I see, sounds good then
